### PR TITLE
scoreboard_func

### PR DESCRIPTION
**Pull request by breeswish**
- Rename "contest status" to "contest scoreboard" to avoid confusing with another "contest status" (live, done, ...)
- Implement scoreboard_func

### CODE REVIEW

File: vj4/handler/contest.py [hunk 1]
```diff
@@ -56,6 +56,17 @@ def status_text(self, tdoc):
   def can_show(self, tdoc):
     return contest.RULES[tdoc['rule']].show_func(tdoc, self.now)
 
+  async def get_scoreboard(self, tid: objectid.ObjectId, is_export: bool=False):
+    tdoc, tsdocs = await contest.get_and_list_status(self.domain_id, tid)
+    if (not contest.RULES[tdoc['rule']].show_func(tdoc, self.now)
+        and not self.has_perm(builtin.PERM_VIEW_CONTEST_HIDDEN_SCOREBOARD)):
+      raise error.ContestScoreboardHiddenError()
```
Comment (twd2) [inline vj4/handler/contest.py]: Arguments

File: vj4/model/adaptor/contest.py [hunk 2]
```diff
@@ -160,7 +231,7 @@ def _get_status_journal(tsdoc):
   """This method returns None when the modification has been superseded by a parallel operation."""
   tdoc = await document.get(domain_id, document.TYPE_CONTEST, tid)
   if pid not in tdoc['pids']:
-    raise error.ValidationError('pid')
+    return None
```
Comment (twd2) [inline vj4/model/adaptor/contest.py]: ???

File: vj4/test/test_contest.py [hunk 3]
```diff
@@ -22,7 +22,8 @@ def time(jdoc):
 
 
 NOW = datetime.datetime.utcnow().replace(microsecond=0)
-TDOC = {'pids': [777, 778, 779], 'begin_at': NOW}
+TDOC = {'pids': [777, 778, 779],
+        'begin_at': NOW}
```
Comment (twd2) [inline vj4/test/test_contest.py]: ???

File: vj4/test/test_contest.py [hunk 4]
```diff
@@ -167,12 +169,6 @@ def tearDown(self):
       await contest.attend(DOMAIN_ID_DUMMY, self.tid, ATTEND_UID)
 
   @base.wrap_coro
-  async def test_update_status_wrong_pid(self):
```
Comment (twd2) [inline vj4/test/test_contest.py]: ???

File: vj4/model/adaptor/contest.py [hunk 5]
```diff
@@ -44,15 +44,86 @@ def time(jdoc):
           'detail': detail}
 
 
-def _oi_equ_func(a, b):
-  return a.get('score', 0) == b.get('score', 0)
+def _oi_status(is_export, translate_func, tdoc, ranked_tsdocs, udict, pdict):
+  columns = []
+  columns.append({'kind': 'rank', 'value': translate_func('Rank')})
```
Comment (twd2) [inline vj4/model/adaptor/contest.py]: Rename `kind` to `type`?

**Comment by twd2** [comment]
Why does wrong pid no longer throw an error?

File: vj4/test/test_contest.py (hunk 3 above)
Comment (breeswish) [reply vj4/test/test_contest.py]: See https://github.com/vijos/vj4/pull/331/files#diff-15cf4000cd97361436e8ffa087659994R26

**Pull request merged**